tidb-binlog/pump_client/client.go

```diff
@@ -0,0 +1,431 @@
+	return nil
+}
+
+// WriteBinlog writes binlog to a situable pump.
+func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
+	pump := c.Selector.Select(binlog)
+	if pump == nil {
+		return ErrNoAvaliablePump
+	}
+	log.Debugf("[pumps client] write binlog choose pump %s", pump.NodeID)
+
+	commitData, err := binlog.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	req := &pb.WriteBinlogReq{ClusterID: c.ClusterID, Payload: commitData}
+
+	// Retry many times because we may raise CRITICAL error here.
+	for i := 0; i < c.RetryTime; i++ {
+		if pump == nil {
+			return ErrNoAvaliablePump
+		}
+
+		resp, err := c.writeBinlog(req, pump)
+		if err == nil && resp.Errmsg != "" {
+			err = errors.New(resp.Errmsg)
+		}
+		if err == nil {
+			return nil
+		}
+
+		log.Errorf("[pumps client] write binlog error %v", err)
+		if isCriticalError(err) {
+			return err
+		}
+
+		// every pump can retry 5 times, if retry 5 times and still failed, set this pump unavaliable, and choose a new pump.
+		if (i+1)%5 == 0 {
+			c.Pumps.Lock()
+			c.setPumpAvaliable(pump, false)
+			c.Pumps.Unlock()
+			pump = c.Selector.Next(pump, binlog, i/5+1)
+			log.Debugf("[pumps client] avaliable pumps: %v, write binlog choose pump %v", c.Pumps.AvaliablePumps, pump)
+		}
+		time.Sleep(RetryInterval)
+	}
+
+	return ErrWriteBinlog
+}
+
+func (c *PumpsClient) writeBinlog(req *pb.WriteBinlogReq, pump *PumpStatus) (*pb.WriteBinlogResp, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.BinlogWriteTimeout)
+	resp, err := pump.Client.WriteBinlog(ctx, req)
+	cancel()
+
+	return resp, err
+}
+
+// setPumpAvaliable set pump's isAvaliable, and modify NeedCheckPumps or AvaliablePumps.
+func (c *PumpsClient) setPumpAvaliable(pump *PumpStatus, avaliable bool) {
+	pump.IsAvaliable = avaliable
+	if avaliable {
+		err := pump.createGrpcClient()
+		if err != nil {
+			log.Errorf("[pumps client] create grpc client for pump %s failed, error: %v", pump.NodeID, err)
+			pump.IsAvaliable = false
+			return
+		}
+
+		delete(c.Pumps.NeedCheckPumps, pump.NodeID)
+		c.Pumps.AvaliablePumps[pump.NodeID] = pump
+
+	} else {
+		delete(c.Pumps.AvaliablePumps, pump.NodeID)
+		c.Pumps.NeedCheckPumps[pump.NodeID] = pump
+	}
+
+	c.Selector.SetPumps(copyPumps(c.Pumps.AvaliablePumps))
+}
+
+// addPump add a new pump.
+func (c *PumpsClient) addPump(pump *PumpStatus, updateSelector bool) {
+	if pump.State == node.Online {
+		c.Pumps.AvaliablePumps[pump.NodeID] = pump
+	} else {
+		c.Pumps.NeedCheckPumps[pump.NodeID] = pump
+	}
+	c.Pumps.Pumps[pump.NodeID] = pump
+
+	if updateSelector {
+		c.Selector.SetPumps(copyPumps(c.Pumps.AvaliablePumps))
+	}
+}
+
+// updatePump update pump's status.
+func (c *PumpsClient) updatePump(status *node.Status) {
+	c.Pumps.Pumps[status.NodeID].updateStatus(status)
+}
+
+// removePump removes a pump.
+func (c *PumpsClient) removePump(nodeID string) {
+	delete(c.Pumps.Pumps, nodeID)
+	delete(c.Pumps.NeedCheckPumps, nodeID)
+	delete(c.Pumps.AvaliablePumps, nodeID)
```

Comment:
need to close connection for this pump?

Reply:
maybe the grpc connection will auto close, I will check it

Reply:
any conclusion?

Reply:
we need close grpc connection explicitly